feat: added source generator benchmark

TimothyMakkison opened the pull request:
## Added benchmark to measure Mapperly source generator performance.

### Description

Adapted code from [Mediator](https://github.com/martinothamar/Mediator/tree/main/benchmarks/Mediator.Benchmarks/SourceGenerator) to benchmark Mapperly source generator performance. This benchmark uses the sample project although a different more complex file could be used instead.

odaiwa commented on benchmarks/Riok.Mapperly.Benchmarks/SourceGeneratorBenchmark.cs:
```diff
@@ -0,0 +1,80 @@
+
+        var projectFile = Path.Combine(currentDirectory, "../../samples/Riok.Mapperly.Sample/Riok.Mapperly.Sample.csproj");
+        if (!File.Exists(projectFile))
+            throw new Exception("Project doesnt exist");
```
```suggestion
            throw new Exception("Project doesn't exist");
```

latonz commented on benchmarks/Riok.Mapperly.Benchmarks/SourceGeneratorBenchmark.cs:
```diff
@@ -0,0 +1,80 @@
+
+[MemoryDiagnoser]
+[InProcess]
+//[EventPipeProfiler(EventPipeProfile.CpuSampling)]
+//[EtwProfiler]
+//[DisassemblyDiagnoser]
```
Shouldn't this commented code be removed before going to main? Or add comments explaining why it is commented and what activating these attributes does.

latonz commented on benchmarks/Riok.Mapperly.Benchmarks/SourceGeneratorBenchmark.cs:
```diff
@@ -0,0 +1,80 @@
+            ConsoleLogger.Default.WriteLineError("-------------------------");
+        };
+
+        var projectFile = Path.Combine(currentDirectory, "../../samples/Riok.Mapperly.Sample/Riok.Mapperly.Sample.csproj");
```
Extract this magic string into a named constant.

latonz reviewed (changes requested):
Thank you for this contribution.
* adding a `launchSettings.json` would be helpful as several IDE's load the launchsettings as run configurations into the IDE
* I think a benchmark of the generated mapping at runtime would be helpful too, wdyt? (probably benchmark `Riok.Mapperly.IntegrationTests.Mapper.StaticTestMapper.MapToDto` as it contains a bit of everything?)
* To be discussed: would it make sense to continuously monitor the performance? Eg. via https://github.com/benchmark-action/github-action-benchmark? (I never used that before, just found it by doing a quick google search)

latonz commented on benchmarks/Riok.Mapperly.Benchmarks/SourceGeneratorBenchmark.cs:
```diff
@@ -0,0 +1,80 @@
+        Project? project = null;
+        try
+        {
+            ConsoleLogger.Default.WriteLine("Loading project!!!");
```
Nit: I don't think these exclamation marks at the end add any value. These messages don't seem to important to me, that they would need a high attention.

TimothyMakkison commented:
> * adding a `launchSettings.json` would be helpful as several IDE's load the launchsettings as run configurations into the IDE

What should be inside this file? afaik this benchmark doesn't use configuration for args.

> * To be discussed: I think a benchmark of the generated mapping at runtime would be helpful too, wdyt? (probably benchmark `Riok.Mapperly.IntegrationTests.Mapper.StaticTestMapper.MapToDto` as it contains a bit of everything?)

I thought it would be really useful although it probably won't be worth `github-action-benchmark` on this. I also thought we could add a benchmark comparing Mapperly to other .NET benchmarking libraries.

> * To be discussed: would it make sense to continuously monitor the performance? Eg. via https://github.com/benchmark-action/github-action-benchmark? (I never used that before, just found it by doing a quick google search)

Good idea, I think I've use this in a different project. I'll try to add a .yml file.

latonz commented on benchmarks/Riok.Mapperly.Benchmarks/MapperComparison/BenchmarkAll.cs:
```diff
@@ -0,0 +1,80 @@
+    private readonly Foo _fooInstance = TestHelper.SetupFooInstance();
+    private readonly Customer _customerInstance = TestHelper.SetupCustomerInstance();
+
+    [Params(100_000)] //, 1_000_000)]
```
Remove commented code.

latonz commented on benchmarks/Riok.Mapperly.Benchmarks/MapperComparison/Models/Customer.cs:
```diff
@@ -0,0 +1,55 @@
+﻿namespace Riok.Mapperly.Benchmarks.MapperComparison.Models;
+
+public class Address
```
This code base usually creates one file per type.

latonz commented on benchmarks/Riok.Mapperly.Benchmarks/SourceGeneratorBenchmark.cs:
```diff
@@ -0,0 +1,88 @@
+    {
+        var currentDirectory = CallerFilePath();
+
+        ConsoleLogger.Default.WriteLine($"Starting at {currentDirectory}");
```
I don't think this adds any benefit.

latonz commented on .github/workflows/sourcegenerator-benchmark.yml:
```diff
@@ -0,0 +1,50 @@
+        with:
+          name: Benchmark.Net Benchmark
+          tool: "benchmarkdotnet"
+          output-file-path: examples/benchmarkdotnet/BenchmarkDotNet.Artifacts/results/Sample.Benchmarks-report-full-compressed.json
```
Is this really correct?

latonz commented on .github/workflows/sourcegenerator-benchmark.yml:
```diff
@@ -0,0 +1,50 @@
+          alert-threshold: "200%"
+          comment-on-alert: true
+          fail-on-alert: true
+          alert-comment-cc-users: "@ktrz"
```
I don't think we need to cc users, as this project still has only a handful of maintainers and each PR is reviewed anyway.

latonz commented on .github/workflows/sourcegenerator-benchmark.yml:
```diff
@@ -0,0 +1,50 @@
+          fail-on-alert: true
+          alert-comment-cc-users: "@ktrz"
+
+      - name: Store benchmark result - separate results repo
```
Isn't this a duplicate of the above step?

latonz reviewed (changes requested):
> What should be inside this file? afaik this benchmark doesn't use configuration for args.

Rider shows nice ready-to run configurations if a `launchSettings.json` exists, can't tell what VS does. Therefore it helps at least for Rider users.

TimothyMakkison commented on benchmarks/Riok.Mapperly.Benchmarks/Riok.Mapperly.Benchmarks.csproj:
```diff
@@ -0,0 +1,35 @@
+    <PackageReference Include="Microsoft.CodeAnalysis.Common" Version="4.5.0" PrivateAssets="all" />
+    <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.4" PrivateAssets="all" />
+
+    <PackageReference Include="ExpressionDebugger" Version="2.2.1" />
+    <PackageReference Include="ExpressionTranslator" Version="2.5.0" />
+    <PackageReference Include="AutoMapper" Version="10.1.1" />
+    <PackageReference Include="FastExpressionCompiler" Version="3.2.1" />
```
Okay I'll remove the comparisons 👍  tbh they were really handy, I found two bugs!

TimothyMakkison commented:
Thanks for your help, I've updated the benchmark, removed the comparisons and modified the action file. 

tbh this is my second time writing a github action and It looks like it will someone with admin level control to setup. Shall I delete the benchmark.yml file and let you or someone with more experience set it up?

TimothyMakkison commented on .github/workflows/sourcegenerator-benchmark.yml:
```diff
@@ -0,0 +1,50 @@
+on:
+  push:
+    branches:
+      - main
```
It will take a while to un eahc time but I've changed it to:
```yml
workflow_call:
pull_request:
  branches:
    - "**"
push:
  branches:
    - main
```

TimothyMakkison replied:
See also https://github.com/benchmark-action/github-action-benchmark#caveats

TimothyMakkison commented:
> You now removed the benchmark of the generated code of Mapperly too, did you? I think it would make sense to keep that one.

Readded the benchamarks from Mapster but removed comparisons to the other mappers,

> To be discussed: I think a benchmark of the generated mapping at runtime would be helpful too, wdyt? (probably benchmark Riok.Mapperly.IntegrationTests.Mapper.StaticTestMapper.MapToDto as it contains a bit of everything?)

Added additional benchmark to `SourceGeneratorBenchmarks` where the time to generate `Riok.Mapperly.IntegrationTests` is measured. I tried adding `StaticTestMapper.MapToDto` as a benchmark but I kept getting a metadata error about not being able to access .NET 7 dll. I suspect the multi targeting logic in the integration tests causes this.

![image](https://github.com/riok/mapperly/assets/49349513/13e35442-cd6f-45e0-9327-f751d1ff8480)

TimothyMakkison commented on benchmarks/Riok.Mapperly.Benchmarks/MappingBenchmarks.cs:
```diff
@@ -0,0 +1,118 @@
+    private readonly Foo _foo;
+    private readonly Customer _customer;
+
+    public MappingBenchmarks()
```
mb I thought you requested that I readd the mapster comparison benchmarks but without the benchmarks for the other libraries.  

> another Mapper around which would need updates with new features, etc.?

Im not sure what you mean by this, why would the current `MappingBenchmarks` need updates?

>  You mention a dll error, do you have more information on that? Probably I'll give it a try in the upcoming days.

I'll see if I can find my stash for this 👍. It would be nice to have a large complex mapper to track.

latonz replied:
> Im not sure what you mean by this, why would the current MappingBenchmarks need updates?

If a new feature is implemented (which should usually then also be in the benchmark and integration tests), `Foo`, `Customer`, `CustomerDto`, `MapperlyMapper` and `MappingBenchmarks` probably all need to be updated whereas if the integration test mappers could be reused changes would only have to be done once for a new feature.

TimothyMakkison commented:
I've removed the Mapster benchmarks and adapted the integration tests into `MapComplexToDto`, `MapComplexToExistingDto`, `MapSimpleDeepClone` and `MapComplexDeepClone`. 

>  You mention a dll error, do you have more information on that? Probably I'll give it a try in the upcoming days.

I wrote and ran these tests in Rider, I suspect this is a Visual Studio issue, I can't even recreate the dll error because VS won't build any of the recent branches. I keep getting the following instead:
`Generator 'MapperGenerator' failed to generate source. Exception was of type 'InvalidOperationException' with message 'Type Riok.Mapperly.Abstractions.EnumMappingStrategy, Riok.Mapperly.Abstractions not found`
Probably caused by installing the latest version of VS or .NET 🤷 

> If a new feature is implemented (which should usually then also be in the benchmark and integration tests), Foo, Customer, CustomerDto, MapperlyMapper and MappingBenchmarks probably all need to be updated whereas if the integration test mappers could be reused changes would only have to be done once for a new feature.

Gotcha 👍

TimothyMakkison commented:
This PR should be finished, it stopped working at one point so I added some packages to fix this I hope this isn't an issue.

latonz commented on benchmarks/Riok.Mapperly.Benchmarks/Riok.Mapperly.Benchmarks.csproj:
```diff
@@ -0,0 +1,55 @@
+    <ProjectReference Include="..\..\test\Riok.Mapperly.IntegrationTests\Riok.Mapperly.IntegrationTests.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <None Remove="BenchmarkDotNet.Artifacts\Riok.Mapperly.Benchmarks.SourceGeneratorBenchmarks-20230529-231345.log" />
+    <None Remove="BenchmarkDotNet.Artifacts\Riok.Mapperly.Benchmarks.SourceGeneratorBenchmarks-20230529-230725.log" />
+    <None Remove="BenchmarkDotNet.Artifacts\Riok.Mapperly.Benchmarks.SourceGeneratorBenchmarks-20230529-230102.log" />
+    <None Remove="BenchmarkDotNet.Artifacts\Riok.Mapperly.Benchmarks.SourceGeneratorBenchmarks-20230529-225025.log" />
+    <None Remove="BenchmarkDotNet.Artifacts\Riok.Mapperly.Benchmarks.SourceGeneratorBenchmarks-20230529-224916.log" />
+    <None Remove="BenchmarkDotNet.Artifacts\results\Riok.Mapperly.Benchmarks.SourceGeneratorBenchmarks-report.html" />
+    <None Remove="BenchmarkDotNet.Artifacts\results\Riok.Mapperly.Benchmarks.SourceGeneratorBenchmarks-report-github.md" />
+    <None Remove="BenchmarkDotNet.Artifacts\Riok.Mapperly.Benchmarks.SourceGeneratorBenchmarks-20230530-115845.log" />
+    <None Remove="BenchmarkDotNet.Artifacts\Riok.Mapperly.Benchmarks.SourceGeneratorBenchmarks-20230530-115624.log" />
+    <None Remove="BenchmarkDotNet.Artifacts\Riok.Mapperly.Benchmarks.SourceGeneratorBenchmarks-20230530-113524.log" />
+    <None Remove="BenchmarkDotNet.Artifacts\Riok.Mapperly.Benchmarks.SourceGeneratorBenchmarks-20230530-113300.log" />
+    <None Remove="BenchmarkDotNet.Artifacts\Riok.Mapperly.Benchmarks.SourceGeneratorBenchmarks-20230529-231934.log" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Folder Include="BenchmarkDotNet.Artifacts\results\" />
+  </ItemGroup>
```
Is this really needed?

latonz commented:
@TimothyMakkison could you rebase this? I'll merge it afterwards.
